### ISSUE TITLE
Redesign CodeSystem & Concept mechanisms

### ISSUE THREAD

**Issue by EvanMachusak** [FirelyTeam/firely-cql-sdk#519]
When we translate `codesystem` in the expression compiler (and subsequently in our C# code), we type it as an array of Code types, which is resolved by default as a `CqlCode` - subsequently, codesystems are expressed as `CqlCode[]`.  This is also true of concepts.

This poses a problem because the `CodeSystemType` is resolved to `CqlCodeSystem`.

Thus, if we are given an expression in CQL whose type is reported to be `"{urn:hl7-org:elm-types:r1}CodeSystem"` as would be the case for `CodeSystemRef` nodes, and we attempt to resolve this type through normal means (e.g., the `TypeResolver`), we will end up getting `typeof(CqlCodeSystem)`.  But this will not match the type of the `Expression` that is stored in our `DefinitionDictionary` for the corresponding `CodeSystemDef`, which is `CqlCode[]`.

The rationale for using `CqlCode[]` for both systems and concepts was to make it easy to support the [`in` operator](https://cql.hl7.org/09-b-cqlreference.html#in-codesystem).

However, this is fundamentally wrong.  Instead, codesystems, concepts, and valuesets, all being vocabulary for which the `in` operator is defined, should all behave similarly.  We do not represent `valueset` as `CqlCode[]`.

This requires refactoring a significant amount of code - specifically, large swaths of support code for managing code systems and their codes is implemented in the `ExpressionBuilderContext` that can be simplified.  These functions should probably not exist:

```
bool TryGetCodesByCodeSystemName(string codeSystemName, [NotNullWhen(true)] out List<CqlCode>? codes);
bool TryGetCodeSystemName(CodeSystemRef codeSystemRef, [NotNullWhen(true)] out string? url);
bool TryGetCode(CodeRef codeRef, [NotNullWhen(true)] out CqlCode? systemCode);
void AddCode(CodeDef codeDef, CqlCode cqlCode);
```

Instead, the `CqlConcept` and `CqlCodeSystem` should probably be created as proper static members of their resulting classes.  For example, a library like this:

```
library Test version '1.0.0'

codesystem "MyCodeSystem": 'http://hl7.org/my-code-system'
code "my-code": 'my-code' from "MyCodeSystem"
```

Should probably be expressed as:
```
class Test {
public static readonly CqlCode my_code = new CqlCode("my-code", "http://hl7.org/my-code-system");
public static readonly CqlCodeSystem MyCodeSystem = new CqlCodeSystem([my_code]);
```

What merits further discussion is whether, like value sets, code systems could be provided at runtime.  For example, what should this code do:

```
library Test version '1.0.0'

using FHIR version '4.0.1'

include FHIRHelpers version '4.0.1' called FHIRHelpers

codesystem "icd10": 'http://icd10canonicalurl'
code "Sucked into jet engine": 'V97.33XD' from "icd10"

define "true for sure": "Sucked into jet engine" in "icd10"
define "true or false?": Code { code: 'R46.1', system: 'http://icd10canonicalurl' } in "icd10"
define "conditions with ICD 10 codes": [Condition] c where c.code in "icd10"
```

If we constrain the `in` operator to only consider explicitly defined code systems, then `"true or false?"` would unintuitively return false.  Likewise, `"conditions with ICD 10 codes"` would also only be able to identify conditions with code `V97.33XD`.

By spec, all three of these defines should return true.  But, it is not our SDK's concern to supply complete code systems so these operators will work - particularly because some code systems are proprietary and cannot be distributed.  Unlike value sets, it may not be adequate to ask the users to supply a `CodeSystem` resource to the `CqlContext` that contains all of ICD10.  For `CodeSystem` and `Concept`s, we should probably use the interface approach, like we do with comparers and converters.

For example, if we allowed users to provide an `ICodeSystemResolver` which is keyed by canonical URL. then users of the SDK have the choice of whether to consult a complete code system catalogue, or to use a rules-based approach (e.g., does the code match the regex formulation of ICD10?).

This is a significant feature set and should be deferred past 2.0.

**Comment by ewoutkramer** [comment]
This is indeed what `ITerminologyService` (or one of its subinterfaces) does. The "in" operator should also not necessarily be implemented by checking full lists of code, as implicitly defined valuesets cannot be directly queried like this. Some naive implentations in the SDK expand the valueset and then lookup the code, but for larger valuesets being able to optimize lookups is going to be important, so hiding the details behind such an interface is a good idea.

**Comment by baseTwo** [comment]
Still the right ticket, and still substantially open. Assessed item by item against `develop` @ `3810aa572`.

### Done since filing

The **emission** half of the proposal has landed, essentially as written in the issue body. Code systems and concepts are no longer typed as `CqlCode[]`; they are created as proper static members of the generated class:

`Cql/CoreTests/CSharp/TestCodeSystemInclude-1.0.0.g.cs:23,31-33`
```csharp
private static readonly CqlCode _Fulfill_Task = new CqlCode("fulfill", "http://hl7.org/fhir/CodeSystem/task-code");
...
private static readonly CqlCodeSystem _TaskCodeSystem =
  new CqlCodeSystem("http://hl7.org/fhir/CodeSystem/task-code", null, [_Fulfill_Task]);
```
and for concepts, `Cql/CoreTests/CSharp/ConceptDefTest-1.0.0.g.cs:47-70` (`new CqlConcept([...], null)`).

Behind that: `CqlCodeSystemDefinition`/`CqlConceptDefinition`/`CqlCodeDefinition` in `Cql/Cql.Compiler/CqlDefinition.cs:81`, `:93`, `:107` (`CqlCodeSystemDefinition.ReturnType => typeof(CqlCodeSystem)`, `:87`), built in `Cql/Cql.Compiler/CodeBuilderContext.LibraryDefs.cs:23-85`. A code system with no declared codes no longer throws (`:29-30` defaults to `[]`), which was #513's original `KeyNotFound`.

### Not done — and the core defect is now live rather than latent

**1. `CodeSystemRef` was never updated to match.** `Cql/Cql.Compiler/CodeBuilderContext.Refs.cs:103-111` still types the reference as an array of codes:
```csharp
var type = _typeResolver.CodeType.MakeArrayType();     // => CqlCode[]
var definitionCallExpression = InvokeDefinitionThroughRuntimeContext(codeSystemRef.name, codeSystemRef.libraryName, type);
```
while `BaseTypeResolver.CodeSystemType` is `typeof(CqlCodeSystem)` (`Cql/Cql.Runtime/Runtime/BaseTypeResolver.cs:81`). The mismatch this issue described is therefore still there, just with the two sides swapped. I built the current tree and compiled both examples from the issue body:

```
"Sucked into jet engine" in "icd10"
  -> CqlException<ExpressionBuildingError>: Cannot convert Hl7.Cql.Primitives.CqlCode[] to Hl7.Cql.Primitives.CqlCodeSystem.
     (ChangeType @ Cql/Cql.Compiler/CodeBuilderContext.Operators.cs:858, via TranslateElement/ConvertToResultType @ CodeBuilderContext.cs:223)
```
Same error for `"cs" is CodeSystem` (#513). Nothing in the test corpus covers a `CodeSystemRef` in expression position — the only `CodeSystemRef` nodes in any checked-in ELM (`Cql/CoreTests/Input/ELM/Test/QueriesTest-1.0.0.json`) are inside `codeDef.codeSystem`, which is resolved via `TryGetCodeSystemName` and never reaches `Refs.cs:103`.

**2. `in (Codesystem)` is entirely unimplemented.** `Cql/Cql.Runtime/Operators/ICqlOperators.cs` contains **zero** `CodeSystem` members — only the `*InValueSet` family (`:153-165`, `:545-546`). The spec requires six overloads (`spec/cql/condensed/09-b-cqlreference.md:6176-6212`: `String`, `Code`, `Concept`, `List<Code>`, `List<String>`, `List<Concept>` against a `CodeSystemRef`), including "if the code system reference cannot be resolved, a run-time error is thrown".

**3. The helper functions this issue wanted removed all still exist**, in the renamed context class (`ExpressionBuilderContext` -> `LibraryCodeBuilderContext` after the typed-IR port, commit `f7a3fa180`):
- `TryGetCodesByCodeSystemName` — `Cql/Cql.Compiler/LibraryCodeBuilderContext.LibraryDefs.cs:99`
- `TryGetCode` — `:126`
- `AddCode` — `:134`
- `TryGetCodeSystemName` — `:187`

**4. No runtime code-system resolution.** There is no `ICodeSystemResolver`, and no `ITerminologyService` (@ewoutkramer's suggestion above): `grep -rn "ICodeSystemResolver\|ITerminologyService" --include=*.cs Cql/` returns nothing. `CqlCodeSystem` still carries a fixed local `IReadOnlyList<CqlCode> codes` (`Cql/Cql.Abstractions/Primitives/CqlCodeSystem.cs:25`), so even once (1) and (2) are fixed the `"true or false?"` and `"conditions with ICD 10 codes"` examples in this issue body would still answer from the locally declared codes only. Worth noting the spec agrees with the interface direction: "The CodeSystem type in CQL is *not* the definition of a code system, it is a reference to an externally defined code system" (`spec/cql/condensed/09-b-cqlreference.md:130`).

### Relationship to other tickets

- **#49 ("CodeSystem needs the same treatment as ValueSets", 2023)** is substantially the same work, and **this issue is the better-stated version** — it has the design analysis, the `CqlCode[]`-vs-`CqlCodeSystem` diagnosis, and the resolver proposal. #49 contributes two details worth folding in here rather than tracking separately: a `CodeSystemFacade` analogous to `ValueSetFacade`, and enumerating a code system with a query (`from CS code return code`). Recommend closing #49 as a duplicate of this issue after copying those two points across.
- **#513** is the concrete regression case for the `is CodeSystem` half and is correctly blocked on this issue.

### Suggested split, if this is picked up

The type-alignment fix (1) plus the operator surface (2) are a bounded, testable unit and would close #513 and #49's first example. The resolver/facade feature (4) is the "significant feature set" the issue says to defer, and deserves its own ticket once (1)+(2) land. Both change emitted C#, so both need a `LibrarySetCSharpCodeGenerator.GeneratorToolVersion` bump and a `*.g.cs` regeneration.

Confidence: high — every claim above was checked against the current tree, and the two failing examples were reproduced against a build of it.
